python/tank/util/shotgun/publish_resolve.py

```diff
@@ -0,0 +1,469 @@
+    #  'id': 25826,
+    #  'link_type': 'local',
+    #  'local_path': '/Users/foo.png',
+    #  'local_path_linux': None,
+    #  'local_path_mac': '/Users/foo.png',
+    #  'local_path_windows': None,
+    #  'local_storage': {'id': 39,
+    #                    'name': 'home',
+    #                    'type': 'LocalStorage'},
+    #  'name': 'foo.png',
+    #  'type': 'Attachment',
+    #  'url': 'file:///Users/foo.png'}
+
+    log.debug("Attempting to resolve local file link attachment data "
+              "into a local path: %s" % pprint.pformat(attachment_data))
+
+    # see if we have a path for this storage
+    local_path = attachment_data.get("local_path")
+    if local_path is None:
+        raise PublishPathNotDefinedError(
+            "Publish '%s' does not have a path defined "
+            "for this platform. Windows Path: '%s', Mac Path: '%s', "
+            "Linux Path: '%s'" % (
+                attachment_data["name"],
+                attachment_data["local_path_windows"],
+                attachment_data["local_path_mac"],
+                attachment_data["local_path_linux"],
+            )
+        )
+
+    # normalize
+    local_path = ShotgunPath.from_current_os_path(local_path).current_os
+
+    # check override env var
+    storage_name = attachment_data["local_storage"]["name"].upper()
+    storage_id = attachment_data["local_storage"]["id"]
+    os_name = {"win32": "WINDOWS", "linux2": "LINUX", "darwin": "MAC"}[sys.platform]
+    env_var_name = "SHOTGUN_PATH_%s_%s" % (os_name, storage_name)
+    log.debug("Looking for override env var '%s'" % env_var_name)
+
+    if env_var_name in os.environ:
+        # we have an override
+        override_root = os.environ[env_var_name]
+        # normalize path
+        override_root = ShotgunPath.from_current_os_path(override_root).current_os
+        log.debug("Applying override '%s' to path '%s'" % (override_root, local_path))
+
+        # get local storages to get path prefix to replace
+        storage = [s for s in get_cached_local_storages(tk) if s['id'] == storage_id][0]
+        # normalize and compare
+        local_storage_root = ShotgunPath.from_shotgun_dict(storage).current_os
+        # apply modification
+        local_path = override_root + local_path[len(local_storage_root):]
+        log.debug("Overridden path: '%s'" % local_path)
+
+    log.debug("Resolved local file link: '%s'" % local_path)
+    return local_path
+
+
+def __resolve_url_link(tk, attachment_data):
+    """
+    Resolves the a given url attachment into a local path.
+
+    The supports the resolution of ``file://`` urls into paths. Such urls
+    are not multi platform and Shotgun local storages and environment variables
+    will be used to try to resolve such paths in case of ambiguity.
+
+    - First, local storage settings will be downloaded from Shotgun and added to
+      a path translation table.
+
+    - Next, any environment variables on the form ``SHOTGUN_PATH_WINDOWS|MAC|LINUX``
```

Comment:
Same comment about `on` vs `of`

I'm going to stop mentioning this now, you can do a search and replace if it turns out I'm right. Otherwise carry on and I'll have learned something new about english grammar. :)